### PR TITLE
cni: Add generic-device chaining plugin

### PR DESCRIPTION
**Pull request by liuxu623**
Some cni plugin doesn't use veth, such as [host-device](https://www.cni.dev/plugins/current/main/host-device/).
In TKE(Tencent Kubernetes Engine), for high permormance, we support per pod use a ENI, we just move ENI to Pod NetNS like [host-device](https://www.cni.dev/plugins/current/main/host-device/).
image
We hope use cilium provide Service and NetworkPolicy, so I implement a chining plugin called `generic-device`.
If use this plugin, cilium-agent will attach bpf program to pod netns interface instead of host veth interface.

### CODE REVIEW

File: pkg/datapath/loader/loader.go [hunk 1]
```diff
@@ -288,24 +289,27 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
...
+				return fmt.Errorf("failed to open current netns: %s", err)
 			}
-			return err
 		}
-		defer finalize()
-
-		if ep.RequireEgressProg() {
-			finalize, err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolToEndpoint, dirEgress, "")
+		defer netNS.Close()
+		return netNS.Do(func(_ ns.NetNS) error {
```
Comment (squeed) [inline pkg/datapath/loader/loader.go:306]: This is a big change - moving the BPF program from the host to the container namespace. What sort of effects will this have? What about existing pods on upgrade?

Comment (liuxu623) [reply pkg/datapath/loader/loader.go:306]: If we didn't use `generic-device` chaining plugin before, netNS is host netNS，it does't affect existing pods on upgrade.
It only works for pods that use `generic-device` chaining plugin.

Comment (brb) [reply pkg/datapath/loader/loader.go:306]: Why can't you move the device from the CNI? It's very similar to what we had in the ipvlan mode which was removed by https://github.com/cilium/cilium/pull/20453. Also, I see some similarities with https://github.com/cilium/cilium/pull/14436.

**Comment by gandro** [comment]
Marking as draft due to inactivity. You can make it ready for review again once the feedback has been discussed.

**Pull request closed without merging**